File: scripts/probe-codex-remote.mjs
import { createServer } from "node:http";
import { spawn } from "node:child_process";
import { createRequire } from "node:module";
import { mkdirSync, writeFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { randomUUID } from "node:crypto";
const require = createRequire(
  new URL("../vendor/claw-empire/package.json", import.meta.url),
);
const { WebSocketServer } = require("ws");
const [container, root, codex] = process.argv.slice(2),
  route = "/exec/" + randomUUID();
const isolatedHome = join(root, "mock-controller-home"),
  isolatedCwd = join(root, "mock-controller-work");
mkdirSync(isolatedHome);
mkdirSync(isolatedCwd);
const methods = [],
  requests = [],
  children = new Set();
let count = 0;
const q = (s) => "'" + s.replaceAll("'", "'\\''") + "'";
const payload = `const fs=require('fs');fs.writeFileSync('/work/model-tool.txt','REMOTE_CODEX_TOOL');try{fs.writeFileSync(${JSON.stringify(join(root, "unexpected-host-write"))},'BAD')}catch{};console.log('REMOTE_CODEX_TOOL');`;
const server = createServer(async (req, res) => {
  const chunks = [];
  for await (const chunk of req) chunks.push(chunk);
  let body;
  try {
    body = JSON.parse(Buffer.concat(chunks).toString());
  } catch {
    body = {};
  }
  requests.push({ path: req.url, tools: body.tools });
  if (!req.url.includes("/responses")) {
    res.writeHead(404);
    res.end("{}");
    return;
  }
  count++;
  const id = "fixture-" + count;
  const events = [{ type: "response.created", response: { id } }];
  if (count === 1) {
    const flattened = (body.tools ?? []).flatMap((t) => t.tools ?? [t]);
    const selected = flattened.find((t) =>
      ["exec_command", "shell_command", "shell"].includes(t.name),
    );
    if (!selected) {
      res.writeHead(500);
      res.end("No supported shell tool");
      return;
    }
    const command = "node -e " + q(payload);
    const args =
      selected.name === "exec_command"
        ? {
            cmd: command,
            workdir: "/work",
            yield_time_ms: 1000,
            max_output_tokens: 500,
          }
        : selected.name === "shell_command"
          ? { command, workdir: "/work", timeout_ms: 10000 }
          : {
              command: ["/bin/sh", "-c", command],
              workdir: "/work",
              timeout_ms: 10000,
            };
    events.push({
      type: "response.output_item.done",
      item: {
        type: "function_call",
        call_id: "fixture-tool",
        name: selected.name,
        arguments: JSON.stringify(args),
      },
    });
  } else if (count === 2)
    events.push({
      type: "response.output_item.done",
      item: {
        type: "custom_tool_call",
        call_id: "fixture-patch",
        name: "apply_patch",
        input:
          "*** Begin Patch\n*** Add File: /work/patched.txt\n+REMOTE_CODEX_PATCH\n*** End Patch",
      },
    });
  else if (count === 3)
    events.push({
      type: "response.output_item.done",
      item: {
        type: "function_call",
        call_id: "fixture-local-denial",
        name: "exec_command",
        arguments: JSON.stringify({
          cmd:
            "node -e " +
            q(
              `require('fs').writeFileSync(${JSON.stringify(join(root, "forbidden-local-fallback"))},'BAD')`,
            ),
          workdir: root,
          environment_id: "local",
          yield_time_ms: 1000,
          max_output_tokens: 500,
        }),
      },
    });
  else
    events.push({
      type: "response.output_item.done",
      item: {
        type: "message",
        role: "assistant",
        content: [{ type: "output_text", text: "Fixture finished." }],
      },
    });
  events.push({
    type: "response.completed",
    response: {
      id,
      usage: { input_tokens: 0, output_tokens: 0, total_tokens: 0 },
    },
  });
  res.writeHead(200, {
    "Content-Type": "text/event-stream",
    Connection: "close",
  });
  for (const event of events)
    res.write("data: " + JSON.stringify(event) + "\n\n");
  res.end();
});
const wsserver = new WebSocketServer({ noServer: true });
server.on("upgrade", (req, socket, head) => {
  if (req.url !== route) {
    socket.destroy();
    return;
  }
  wsserver.handleUpgrade(req, socket, head, (ws) =>
    wsserver.emit("connection", ws),
  );
});
wsserver.on("connection", (ws) => {
  const child = spawn(
    "/opt/homebrew/bin/docker",
    [
      "--host",
      "unix://" + process.env.HOME + "/.colima/pazmo-office/docker.sock",
      "exec",
      "-i",
      container,
      "/runner/codex",
      "exec-server",
      "--listen",
      "stdio",
    ],
    { stdio: ["pipe", "pipe", "pipe"] },
  );
  children.add(child);
  child.on("exit", () => {
    children.delete(child);
    ws.close();
  });
  let tail = "";
  child.stdout.on("data", (chunk) => {
    tail += chunk;
    let i;
    while ((i = tail.indexOf("\n")) >= 0) {
      const line = tail.slice(0, i);
      tail = tail.slice(i + 1);
      if (ws.readyState === 1) ws.send(line);
    }
  });
  child.stderr.on("data", () => {});
  ws.on("message", (data) => {
    try {
      const msg = JSON.parse(data.toString());
      if (msg.method) methods.push(msg.method);
    } catch {}
    child.stdin.write(data.toString() + "\n");
  });
  ws.on("close", () => {
    child.stdin.end();
    child.kill("SIGTERM");
  });
});
await new Promise((r) => server.listen(0, "127.0.0.1", r));
const port = server.address().port;
const args = [
  "exec",
  "--ignore-user-config",
  "--ignore-rules",
  "--skip-git-repo-check",
  "--ephemeral",
  "--json",
  "-C",
  isolatedCwd,
  "-s",
  "danger-full-access",
  "-c",
  'approval_policy="never"',
  "-c",
  'model_provider="fixture"',
  "-c",
  `model_providers.fixture={name="Local fixture",base_url="http://127.0.0.1:${port}/v1",wire_api="responses",requires_openai_auth=false,supports_websockets=false}`,
  "-c",
  'web_search="disabled"',
  "--disable",
  "multi_agent",
  "--disable",
  "apps",
  "--disable",
  "shell_snapshot",
  "-m",
  "gpt-5.4",
  "Execute the supplied local fixture tool call and stop.",
];
const child = spawn(codex, args, {
  env: {
    PATH: process.env.PATH,
    HOME: isolatedHome,
    CODEX_HOME: isolatedHome,
    CODEX_EXEC_SERVER_URL: `ws://127.0.0.1:${port}${route}`,
    LANG: "en_US.UTF-8",
  },
  stdio: ["ignore", "pipe", "pipe"],
});
let stdout = "",
  stderr = "";
child.stdout.on("data", (x) => (stdout += x));
child.stderr.on("data", (x) => (stderr += x));
const timer = setTimeout(() => child.kill("SIGTERM"), 45000);
const exit = await new Promise((r) =>
  child.on("exit", (code, signal) => r({ code, signal })),
);
clearTimeout(timer);
for (const ws of wsserver.clients) ws.terminate();
for (const proc of children) proc.kill("SIGTERM");
await new Promise((r) => server.close(r));
wsserver.close();
const report = {
  exit,
  methods,
  requests,
  stdout,
  stderr,
  unexpectedHostWrite: existsSync(join(root, "unexpected-host-write")),
  localFallback: existsSync(join(root, "forbidden-local-fallback")),
};
writeFileSync(
  join(root, "mock-controller-report.json"),
  JSON.stringify(report, null, 2),
);
console.log(
  JSON.stringify({
    exit,
    methods,
    requestCount: count,
    unexpectedHostWrite: report.unexpectedHostWrite,
    report: join(root, "mock-controller-report.json"),
  }),
);

File: scripts/probe-codex-remote.py
"""No model calls or authentication material; disposable existing-VM transport proof."""
import argparse, base64, hashlib, json, os, pathlib, queue, shutil, subprocess, sys, tempfile, threading, time, uuid

D = [shutil.which('docker') or '/opt/homebrew/bin/docker', '--host', f'unix://{pathlib.Path.home()}/.colima/pazmo-office/docker.sock']
IMAGE = 'node@sha256:a9f5f7c91a432850b2a8a7797adf5eadb6c733ceed61167806cee7ea7fbc29df'
parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument('--binary', required=True, help='Verified official Linux arm64 Codex 0.154.0 binary')
parser.add_argument('--mock-controller', action='store_true', help='Also test local Codex 0.155.1 against a fixed localhost model fixture')
options = parser.parse_args()
BINARY = options.binary
with open(BINARY, 'rb') as binary_file:
    binary_hash = hashlib.sha256()
    for chunk in iter(lambda: binary_file.read(1024 * 1024), b''):
        binary_hash.update(chunk)
if binary_hash.hexdigest() != '9b7c1c7abdc26fc3c4f47c77656a8e9121def5483dbae830ef1ee561758448a9':
    raise SystemExit('Refusing unverified exec-server binary')
root = pathlib.Path(tempfile.mkdtemp(prefix='pazmo-exec-server-', dir='/private/tmp'))
name = 'pazmo-exec-probe-' + uuid.uuid4().hex
volume = name + '-bin'
created = []
volume_created = False
child = None
report = {'scope': 'unauthenticated exec-server in disposable isolated VM container', 'liveReady': False,
          'fixtureRoot': str(root), 'events': [], 'checks': [], 'errors': []}
def run(args):
    process = subprocess.Popen(args, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, start_new_session=True)
    try:
        stdout, stderr = process.communicate(timeout=60)
    except subprocess.TimeoutExpired:
        import signal
        os.killpg(process.pid, signal.SIGKILL)
        process.communicate()
        raise
    if process.returncode:
        raise RuntimeError(f'Command failed ({process.returncode}): {args[0]}: {stderr[:1000]}')
    return subprocess.CompletedProcess(args, process.returncode, stdout, stderr)
def check(name, passed):
    report['checks'].append({'name': name, 'passed': bool(passed)})
try:
    if run(D + ['info', '--format', '{{.Name}}']).stdout.strip() != 'colima-pazmo-office':
        raise RuntimeError('Refusing a daemon other than the dedicated VM')
    if options.mock_controller:
        codex = shutil.which('codex')
        if not codex or run([codex, '--version']).stdout.strip() != 'codex-cli 0.155.1':
            raise RuntimeError('Mock controller probe requires observed Codex CLI 0.155.1')
    fake = root / 'private-controller-canary'; fake.write_text('FAKE SECRET ONLY')
    run(D + ['volume', 'create', '--label', 'pazmo.probe=' + name, volume]); volume_created = True
    seed = name + '-seed'
    run(D + ['create', '--name', seed, '--pull', 'never', '--network', 'none', '--mount',
             f'type=volume,src={volume},dst=/runner', IMAGE, 'true']); created.append(seed)
    run(D + ['cp', BINARY, seed + ':/runner/codex'])
    flags = ['create', '--name', name, '--pull', 'never', '--interactive', '--network', 'none',
             '--read-only', '--user', '1000:1000', '--cap-drop', 'ALL', '--security-opt', 'no-new-privileges:true',
             '--pids-limit', '64', '--memory', '512m', '--memory-swap', '512m', '--cpus', '1',
             '--tmpfs', '/work:rw,nosuid,nodev,noexec,size=32m,uid=1000,gid=1000,mode=700',
             '--tmpfs', '/tmp:rw,nosuid,nodev,noexec,size=32m,uid=1000,gid=1000,mode=700',
             '--env', 'HOME=/work', '--env', 'CODEX_HOME=/work/codex-home', '--workdir', '/work',
             '--mount', f'type=volume,src={volume},dst=/runner,readonly,volume-nocopy',
             '--entrypoint', '/runner/codex', IMAGE, 'exec-server', '--listen', 'stdio']
    run(D + flags); created.append(name)
    stderr = open(root / 'stderr.txt', 'w')
    child = subprocess.Popen(D + ['start', '--attach', '--interactive', name], stdin=subprocess.PIPE,
                             stdout=subprocess.PIPE, stderr=stderr, text=True, bufsize=1)
    incoming = queue.Queue()
    def read():
        for line in child.stdout:
            try: incoming.put(json.loads(line))
            except Exception: incoming.put({'unparsed': line[:200]})
        incoming.put({'eof': True})
    threading.Thread(target=read, daemon=True).start()
    counter = 0
    def rpc(method, params):
        global counter
        counter += 1
        child.stdin.write(json.dumps({'id': counter, 'method': method, 'params': params}) + '\n'); child.stdin.flush()
        deadline = time.monotonic() + 15
        while time.monotonic() < deadline:
            event = incoming.get(timeout=max(.1, deadline - time.monotonic()))
            if event.get('id') == counter:
                report['events'].append({'method': method, 'response': event})
                return event
            if event.get('eof'): raise RuntimeError('exec-server EOF')
        raise TimeoutError(method)
    initialized = rpc('initialize', {'clientName': 'pazmo-isolation-probe'})
    check('official-exec-server-initialized', 'result' in initialized)
    child.stdin.write(json.dumps({'method': 'initialized', 'params': {}}) + '\n'); child.stdin.flush()
    payload = """const fs = require('fs');
fs.writeFileSync('/work/result.txt', 'tool execution works');
const r = {write: fs.readFileSync('/work/result.txt','utf8') === 'tool execution works'};
try {fs.readFileSync(%s);r.hostControllerDenied=false;} catch {r.hostControllerDenied=true;}
try {fs.writeFileSync('/runner/codex','corrupt');r.runtimeReadonly=false;} catch {r.runtimeReadonly=true;}
r.authAbsent=!fs.existsSync('/work/codex-home/auth.json');
console.log(JSON.stringify(r));
""" % json.dumps(str(fake))
    started = rpc('process/start', {'processId': 'probe', 'argv': ['node', '-e', payload],
                                  'cwd': 'file:///work', 'env': {}, 'tty': False, 'arg0': None})
    check('remote-process-started', 'result' in started)
    output = ''
    if 'result' in started:
        cursor = None
        for _ in range(20):
            response = rpc('process/read', {'processId': 'probe', 'afterSeq': cursor,
                                           'maxBytes': 65536, 'waitMs': 500})
            result = response.get('result', {})
            output += ''.join(base64.b64decode(c['chunk']).decode() for c in result.get('chunks', []))
            if result.get('chunks'): cursor = result['chunks'][-1]['seq']
            if result.get('closed'):
                check('remote-process-exit-zero', result.get('exitCode') == 0)
                break
        report['processOutput'] = output
        for k,v in json.loads(output).items(): check(k, v)
    fsresult = rpc('fs/readFile', {'path': 'file:///work/result.txt', 'sandbox': None})
    check('remote-filesystem-read', base64.b64decode(fsresult.get('result', {}).get('dataBase64', '')).decode() == 'tool execution works')
    denied = rpc('fs/readFile', {'path': fake.as_uri(), 'sandbox': None})
    check('rpc-host-controller-denied', 'error' in denied)
    inspect = json.loads(run(D + ['inspect', name]).stdout)[0]
    check('no-host-bind-mounts', all(m['Type'] != 'bind' for m in inspect['Mounts']))
    check('container-network-none', inspect['HostConfig']['NetworkMode'] == 'none')
    check('container-capabilities-dropped', inspect['HostConfig']['CapDrop'] == ['ALL'])
    if options.mock_controller:
        mock = run([shutil.which('node') or 'node', str(pathlib.Path(__file__).with_suffix('.mjs')), name, str(root), codex])
        report['mockController'] = json.loads(mock.stdout)
        marker = rpc('fs/readFile', {'path': 'file:///work/model-tool.txt', 'sandbox': None})
        check('mock-controller-tool-executed-remotely', base64.b64decode(marker.get('result', {}).get('dataBase64', '')).decode() == 'REMOTE_CODEX_TOOL')
        check('mock-controller-no-host-write', not (root / 'unexpected-host-write').exists())
        controller = json.loads((root / 'mock-controller-report.json').read_text())
        check('mock-controller-explicit-local-denied', not (root / 'forbidden-local-fallback').exists() and 'unknown turn environment id `local`' in controller['stderr'])
        check('mock-controller-exit-zero', controller['exit'] == {'code': 0, 'signal': None})
        patch = rpc('fs/readFile', {'path': 'file:///work/patched.txt', 'sandbox': None})
        check('mock-controller-patch-executed-remotely', base64.b64decode(patch.get('result', {}).get('dataBase64', '')).decode() == 'REMOTE_CODEX_PATCH\n')
except Exception as error:
    report['errors'].append(str(error))
finally:
    for item in reversed(created):
        try: run(D + ['rm', '--force', item])
        except Exception as error: report['errors'].append('cleanup: ' + str(error))
    if child:
        try: child.wait(timeout=5)
        except subprocess.TimeoutExpired: child.kill(); child.wait()
    if volume_created:
        try: run(D + ['volume', 'rm', volume])
        except Exception as error: report['errors'].append('volume cleanup: ' + str(error))
    (root / 'report.json').write_text(json.dumps(report, indent=2))
    print(json.dumps({'report': str(root / 'report.json'), 'checks': report['checks'], 'errors': report['errors']}))

sys.exit(1 if report['errors'] or not report['checks'] or not all(c['passed'] for c in report['checks']) else 0)
